adding port v1->v2 TEST-0048 (#13)
ScreaMy7/owasp-mastg: tests-beta/android/MASVS-RESILIENCE/MASTG-TEST-0x49.md

Reviewer: Rewrite the Overview, Steps, Observation, and Evaluation sections according to https://github.com/OWASP/mastg/blob/master/.github/instructions/mastg-test.instructions.md

---
platform: android
title: Runtime Use of Xposed/LSPosed Detection Mechanisms
id: MASTG-TEST-0x49
apis: [PackageManager, Method, BufferedReader, Throwable, Thread, File]

> type: [dynamic]

Reviewer: ```suggestion
title: Runtime Use of Xposed/LSPosed Detection Techniques
id: MASTG-TEST-0x49
apis: [PackageManager, Method, BufferedReader, Throwable, Thread, File]
type: [dynamic, hooks]
```

weakness: MASWE-0098
best-practices: [MASTG-BEST-0x49]
profiles: [R]
knowledge: [MASTG-KNOW-0030]
---

## Overview

> Well-known Xposed/LSPosed detection patterns a `PackageManager.getPackageInfo` lookup for known Manager package ids, a reflection-based `Modifier.isNative` tripwire on guaranteed-native methods, a `/proc/self/maps` scan for foreign DEX/APK mappings, and a stack-trace / `/proc/self/task` probe for framework frames and native worker threads provide little resilience against an attacker that controls the device. Each of these signals can be neutralized with a few lines of Frida JavaScript that hooks the underlying Java APIs and returns the values the detection logic expects on a clean device.

Reviewer: ```suggestion
This test verifies whether the app detects Xposed/LSPosed at runtime using common local-environment signatures, such as a `PackageManager.getPackageInfo` lookup for known Manager package ids, a reflection-based `Modifier.isNative` tripwire on guaranteed-native methods, a `/proc/self/maps` scan for foreign APK/DEX mappings, or a stack-trace and `/proc/self/task` walk for framework frames and native worker threads. Because these checks rely entirely on user-space APIs controlled by the attacker, they can be silently disabled by hooking the underlying Java calls to return the values expected on a clean device, leaving the framework and its modules undetected.
```

This test checks whether an app's Xposed/LSPosed detection routines actually resist a runtime API-hooking bypass.

## Steps

1. Install the app on a rooted device with LSPosed active and at least one module scoped to the app.
2. Use @MASTG-TOOL-0001 to attach to the app and test the function that perform Xposed/LSPosed detection.

> 3. Run @MASTG-TECH-0033 to attempt to hook the APIs used by each detection routine to return clean values.

Reviewer: ```suggestion
1. Use @MASTG-TECH-0005 to install the app on a rooted device with LSPosed active and at least one module scoped to the app.
2. Use @MASTG-TECH-0043 to hook the relevant API calls.
3. Exercise the app extensively to trigger as many flows as possible and enter sensitive data wherever you can.
```

## Observation

The output should contain a list of detection routines that the app executed, the APIs they queried, and the value those APIs returned both with the a frida script attached.

## Evaluation

> The test case fails if the app fails to detect Xposed/LSPosed when at least one LSPosed module is loaded into the process.

Reviewer: ```suggestion
The test case fails if the app continues to operate normally after each detection routine receives spoofed clean values from the hooked APIs, indicating that the Xposed/LSPosed detection logic can be neutralized at the API level.
```

3. Trigger the app's Xposed detection function and capture the result.

## Observation

The output should contain a list of detection routines, the APIs they queried (`PackageManager.getPackageInfo`, `Method.modifiers`, `/proc/self/maps`, `Throwable.stackTrace`, `Thread.getAllStackTraces`, `/proc/self/task/<tid>/comm`), and the value those APIs returned with and without the bypass attached. It should also indicate whether the app raised the liability-acceptance dialog when the checks fired.

## Evaluation

The test case fails if **any** of the following hold:

- The app reports "no Xposed detected" while at least one LSPosed module is loaded into the process and a generic Java-side bypass is attached, or
- A check fires but the app does not raise an alert/liability-acceptance dialog to the user before proceeding.

> Note that this Java-side test only covers the API-hooking bypass category. Bypass via app-code patching and via kernel-level syscall interception are out of scope and must be addressed by complementary controls (DEX/native integrity hashing, signing-certificate pinning, and server-side attestation such as the Play Integrity API).

Reviewer: ```suggestion
```
